src/lerobot/processor/pipeline.py

```diff
@@ -893,20 +895,84 @@ def __post_init__(self):
                     f"Step {i} ({type(step).__name__}) must define feature_contract(features) -> dict[str, Any]"
                 )
 
-    def feature_contract(self, initial_features: dict[str, PolicyFeature]) -> dict[str, PolicyFeature]:
+    def dataset_features(self, initial_features: dict[str, PolicyFeature]) -> dict[str, PolicyFeature]:
         """
-        Apply ALL steps in order. Each step must implement
-        feature_contract(features) and return a dict (full or incremental schema).
+        Apply ALL steps in order. Only if a step has a dataset_features method, it will be called.
+        We aggregate the dataset features of all steps.
         """
         features: dict[str, PolicyFeature] = deepcopy(initial_features)
 
         for _, step in enumerate(self.steps):
-            out = step.feature_contract(features)
-            if not isinstance(out, dict):
-                raise TypeError(f"{step.__class__.__name__}.feature_contract must return dict[str, Any]")
-            features = out
+            if hasattr(step, "dataset_features"):
+                out = step.dataset_features(features)
+                if not isinstance(out, dict):
+                    raise TypeError(f"{step.__class__.__name__}.dataset_features must return dict[str, Any]")
+                features = out
         return features
 
+    def aggregate_dataset_features(
+        self,
+        initial_features: dict[str, Any],
+        *,
+        use_videos: bool = True,
```

Comment:
Add an optional `pattern` (or equivalent) to filter the key instead to embed all the logic inside the function. Dependency injection, in that case, is better, because it favors single principle concept and composability.